test/integration/podgc/podgc_test.go

```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
+	}
+	defer testutils.RemovePodFinalizers(testCtx.ClientSet, t, []*v1.Pod{pod})
+	pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while updating pod info: '%v'", err, klog.KObj(pod))
+	}
+
+	// we delete the node to orphan the pod
+	err = cs.CoreV1().Nodes().Delete(testCtx.Ctx, pod.Spec.NodeName, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("Failed to delete node: %v, err: %v", pod.Spec.NodeName, err)
+	}
+
+	err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
+		updatedPod, err := cs.CoreV1().Pods(testCtx.NS.Name).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+		if updatedPod.ObjectMeta.DeletionTimestamp != nil {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("Error '%v' while waiting for the pod '%v' to be deleted", err, klog.KObj(pod))
+	}
+}
+
+// TestTerminatingOnOutOfServiceNode tests deletion pods terminating on out-of-service nodes
+func TestTerminatingOnOutOfServiceNode(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.NodeOutOfServiceVolumeDetach, true)()
+	testCtx := setup(t, "podgc-out-of-service")
+	defer testutils.CleanupTest(t, testCtx)
+	cs := testCtx.ClientSet
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node",
+		},
+		Spec: v1.NodeSpec{},
+		Status: v1.NodeStatus{
+			Conditions: []v1.NodeCondition{
+				{
+					Type:   v1.NodeReady,
+					Status: v1.ConditionFalse,
+				},
+			},
+		},
+	}
+	node, err := cs.CoreV1().Nodes().Create(testCtx.Ctx, node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create node '%v', err: %v", node.Name, err)
+	}
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "testpod",
+			Namespace: testCtx.NS.Name,
+		},
+		Spec: v1.PodSpec{
+			NodeName: node.Name,
+			Containers: []v1.Container{
+				{Name: "foo", Image: "bar"},
+			},
+		},
+	}
+
+	pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
+	}
+
+	// trigger termination of the pod, but with long grace period so that it is not removed immediately
+	err = cs.CoreV1().Pods(testCtx.NS.Name).Delete(testCtx.Ctx, pod.Name, metav1.DeleteOptions{GracePeriodSeconds: pointer.Int64(300)})
+	if err != nil {
+		t.Fatalf("Error: '%v' while deleting pod: '%v'", err, klog.KObj(pod))
+	}
+
+	// get the pod status to make sure the pod is not deleted yet, but it is terminating
+	pod, err = cs.CoreV1().Pods(pod.Namespace).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while checking the pod '%v' is not deleted yet", err, klog.KObj(pod))
+	}
+	if pod.DeletionTimestamp == nil {
```

Comment:
I wonder if this could be flaky and we should add PollImmediate.

Reply:
cleaned up